File: growth/oj/leet_code/algorithms/200-number-of-islands.py
# -*- coing: utf-8 -*-
class UnionFind(object):
    def __init__(self, N):
        self.roots = [0] * N
        for i in range(N):
            self.roots[i] = i

    def find_roots(self, i):
        root = i
        while root != self.roots[root]:
            root = self.roots[root]
        # 路径压缩, 用于提升速度, TODO: 补充一个path compress
        while i != self.roots[i]:
            tmp = self.roots[i]
            self.roots[i] = root
            i = tmp
            pass
        return root

    def connected(self, p, q):
        return self.find_roots(p) == self.find_roots(q)

    def union(self, p, q):
        p_root = self.find_roots(p)
        q_root = self.find_roots(q)
        # p<-q
        self.roots[q_root] = p_root
        #print "union, p:%d, q:%d %s\n" % (p, q, self.roots)

    def count_roots(self):
        cnt = 0
        for i in range(len(self.roots)):
            if i == self.roots[i]:
                cnt += 1
        #print self.roots
        return cnt


class Solution(object):
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        num = 0
        record_land = {}
        for raw in range(len(grid)):
            for col in range(len(grid[raw])):
                if grid[raw][col] == '0':
                    continue
                record_land[(raw, col)] = num
                num += 1
        #print num
        #print record_land
        uf = UnionFind(num)

        for raw in range(len(grid)):
            for col in range(len(grid[raw])):
                if grid[raw][col] == '0':
                    continue
                # handle (raw, col - 1)
                if col - 1  >= 0 and grid[raw][col-1] == '1':
                    uf.union(record_land[(raw, col-1)], record_land[(raw, col)])
                if raw - 1 >= 0 and grid[raw-1][col] == '1':
                    uf.union(record_land[(raw-1, col)], record_land[(raw, col)])

        return uf.count_roots()

File: growth/oj/leet_code/algorithms/200-number-of-islands.cpp
#include<iostream>
#include<vector>

using namespace std;

class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        if (grid.size() == 0) {
            return 0;
        }
        vector<vector<int>> mark(grid.size(), vector<int>(grid[0].size()));
        int cnt = 0;
        for (int x = 0; x < grid.size(); x++)
            for (int y = 0; y < grid[x].size(); y++) {
                if (grid[x][y] == '1' && mark[x][y] == 0) {

                    //cout << "x " << x << "  y " << y << endl;
                    cnt += 1;
                    mark[x][y] = 1;
                    dfs(x, y, grid, mark);
                }
            }

        return cnt;
    }

    void dfs(int x, int y, vector<vector<char>>& grid, vector<vector<int>>& mark) {
        //cout << "in x" << x << " y  " << y << endl;
        vector<int>direct_x{-1,1,0,0};
        vector<int>direct_y{0, 0,1, -1};
        for (int i = 0; i < 4; i++) {
            int next_x = x + direct_x[i];
            int next_y = y + direct_y[i];
            if (next_x >= grid.size() || next_x < 0 || next_y >= grid[0].size() ||  next_y < 0)
                //cout << "1-next_x "  << next_x << "  next_y " << next_y << endl;
                continue;
            if (mark[next_x][next_y] != 0 || grid[next_x][next_y] == '0')
                //cout << "2-next_x "  << next_x << "  next_y " << next_y << endl;
                continue;
            mark[next_x][next_y] = 1;
            dfs(next_x, next_y, grid, mark);
        }
    }
};




int main() {
    Solution s;
    return 0;
}
